### PR TITLE
New module: rectangleGrid

### PR DESCRIPTION
**Pull request by stevage**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Have read [How To Contribute](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#how-to-contribute).
- [x] Run `npm test` at the sub modules where changes have occurred.
- [?] Run `npm run lint` to ensure code style at the turf module level.

`npm run lint` doesn't seem to do anything useful at the moment.

Submitting a new TurfJS Module.

- [x] Overview description of proposed module.

See #1403 and #1251.

- [x] Include JSDocs with a basic example.
- [x] Execute `./scripts/generate-readmes` to create `README.md`.
- [x] Add yourself to **contributors** in `package.json` using "Full Name <@GitHub Username>".

### CODE REVIEW

**Comment by rowanwins** [comment]
Thanks @stevage I'll try and take a look at this over the weekend.

**Comment by stebogit** [comment]
@stevage this package is just an exact copy of `@turf/square-grid`, except the additional parameter for the second dimension of the cell.
Since a square is actually a special case of a rectangle I'd suggest to modify the current `square-grid` package to include the additional param, add a few test for the new features and rename it `@turf/rectangle-grid`.

Then create a new package for `@turf/square-grid` as a simple alias for the specific case:
```js
export default function squareGrid(bbox, cellSize, options) {
    return turf.rectangleGrid(bbox, cellSize, cellSize, options);
}
```

That should save a lot of dev time and avoid redundant code.

@rowanwins @DenisCarriere what do you think?

**Comment by rowanwins** [comment]
You're a smarty pants @stebogit - that does sound like a sensible approach as it would keep the modules in sync if any improvements/fixes need to be made.

PS good to see you :)

**Comment by stevage** [comment]
Yeah, I like that approach - but it definitely doesn't "save a lot of dev time" when the PR is already submitted :)

Anyway I'll have a go.

**Comment by stevage** [comment]
Ok, did I do it right @stebogit ?

File: packages/turf-rectangle-grid/package.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/Turfjs/turf.git"
+  },
+  "keywords": [
+    "turf",
+    "grid",
+    "regular",
+    "cartesian",
+    "grid"
```
Comment (stebogit) [inline packages/turf-rectangle-grid/package.json]: this is duplicated

File: packages/turf-square-grid/index.d.ts [hunk 2]
```diff
@@ -22,9 +21,9 @@ import { BBox, Feature, FeatureCollection, MultiPolygon, Polygon, Properties, Un
  * //addToMap
  * var addToMap = [squareGrid]
  */
-declare function squareGrid<P = Properties>(bbox: BBox, cellSide: number, options?: {
+import { FeatureCollection, Polygon, BBox, Units, Feature, MultiPolygon, Properties } from "@turf/helpers";
```
Comment (stebogit) [inline packages/turf-square-grid/index.d.ts]: the import should go at the very top of the file

File: packages/turf-square-grid/index.ts [hunk 3]
```diff
@@ -29,56 +21,19 @@ import {getType} from "@turf/invariant";
...
-    const cellWidth = xFraction * (east - west);
-    const yFraction = cellSide / (distance([west, south], [west, north], options));
-    const cellHeight = yFraction * (north - south);
-
-    // rows & columns
-    const bboxWidth = (east - west);
-    const bboxHeight = (north - south);
-    const columns = Math.floor(bboxWidth / cellWidth);
-    const rows = Math.floor(bboxHeight / cellHeight);
+import {
```
Comment (stebogit) [inline packages/turf-square-grid/index.ts]: this should go at the top, same as above, just for consistency within the library

File: packages/turf-square-grid/package.json [hunk 4]
```diff
@@ -43,9 +43,6 @@
     "@types/tape": "*"
   },
   "dependencies": {
-    "@turf/boolean-intersects": "6.x",
-    "@turf/distance": "6.x",
-    "@turf/helpers": "6.x",
```
Comment (stebogit) [inline packages/turf-square-grid/package.json]: isn't `@turf/helpers` still required here for types?

File: packages/turf-rectangle-grid/LICENSE [hunk 5]
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
+
+Copyright (c) 2017 TurfJS
```
Comment (stebogit) [inline packages/turf-rectangle-grid/LICENSE]: I guess this should be 2018

File: packages/turf-rectangle-grid/package.json [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    "@turf/truncate": "*",
+    "@std/esm": "*",
+    "benchmark": "*",
+    "rollup": "*",
+    "write-json-file": "*",
+    "load-json-file": "*",
+    "tape": "*"
+  },
+  "dependencies": {
+    "@turf/helpers": "*"
```
Comment (stebogit) [inline packages/turf-rectangle-grid/package.json]: `@turf/boolean-intersects`, `@turf/distance` and `@turf/invariant` are missing

File: packages/turf-rectangle-grid/LICENSE (hunk 5 above)
Comment (stevage) [reply packages/turf-rectangle-grid/LICENSE]: Done

File: packages/turf-rectangle-grid/package.json (hunk 6 above)
Comment (stevage) [reply packages/turf-rectangle-grid/package.json]: Done

**Comment by rowanwins** [comment]
Have just pushed some changes here @stevage & @stebogit to get travis to be happy

- Currently the travis tests fail if the licence date is set to anything other than 2017
- I think the addition to the turf master package was also giving it grief so I've temporarily removed it
  - I need to work out how this part of the process was actually being managed

File: packages/turf-square-grid/package.json (hunk 4 above)
Comment (rowanwins) [reply packages/turf-square-grid/package.json]: Yeah from what I can tell `@turf/helpers` now gets brought in pretty much everywhere to help with the typescript side of things.

File: packages/turf-rectangle-grid/package.json [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    "main.js",
+    "main.es.js"
+  ],
+  "scripts": {
+    "pretest": "rollup -c ../../rollup.config.js",
+    "test": "node -r @std/esm test.js",
+    "posttest": "node -r @std/esm ../../scripts/validate-es5-dependencies.js",
+    "bench": "node -r @std/esm bench.js",
+    "docs": "node ../../scripts/generate-readmes"
+  },
```
Comment (rowanwins) [inline packages/turf-rectangle-grid/package.json]: So I think these scripts need changing to match one of the other typescript modules (eg turf-random) just above it.
- Then I think the main.js and the main.es.js can be deleted and the `@std/esm` part of the package.json

File: packages/turf-square-grid/package.json (hunk 4 above)
Comment (rowanwins) [reply packages/turf-square-grid/package.json]: Same story in this package.json as above - we need to make it align to the TS world view.

File: packages/turf-rectangle-grid/package.json (hunk 7 above)
Comment (stevage) [reply packages/turf-rectangle-grid/package.json]: Ok, done. I'm not quite sure why my generated index.js looks different from the one that was already in this branch. Hopefully not a problem?

**Comment by stevage** [comment]
Just checking in, I *think* I have addressed all the outstanding issues, but want to make sure you're not waiting on me to do something.

**Comment by rowanwins** [comment]
Hi @stevage yep I think it's all good. 

I'll accept the pull request and attempt to publish sometime over the next few nights. The publishing process with [lerna](https://github.com/lerna/lerna) is still a bit mysterious to me so I'm just trying to avoid breaking everything :)

**Comment by stevage** [comment]
Oh awesome 👍

**Comment by adam3039** [comment]
Hey @rowanwins just curious if there had been any headway made on a publish of this latest module. I'm having difficulty bringing in the module from just the github repo alone. It looks like the main turf package in packages/turf doesn't point to the rectangle-grid module at all. Possibly something was missed in the pull request, or I could very well be off my rocker as well. I've tried a few different ways to get it working, but so far only using the official releases from NPM have worked for me.

Thanks for all your help with this.

**Comment by stevage** [comment]
>It looks like the main turf package in packages/turf doesn't point to the rectangle-grid module at all.

That's a problem that has to do with a circular dependency: it seems to be hard to both add the module, and have the main module depend on it, at the same time.

Are you able to make that change locally?

**Comment by yuqianma** [comment]
Currently there's no `rectangleGrid ` in `@turf/turf`. Is it intended?

**Pull request merged**